### ISSUE TITLE
Calling `Vec::extend` with a constant slice yields false-positive failures

### ISSUE THREAD

**Issue by roypat** [model-checking/kani#2656]
I tried this code:

```rust
#[kani::proof]
fn verify_my_test() {
    const MY_CONSTANT: &[u8] = b"Hi";

    let mut my_vec: Vec<u8> = Vec::new();
    my_vec.extend(MY_CONSTANT);
}
```

using the following command line invocation:

```
kani test.rs
```

with Kani version: 0.33.0

I expected to see this happen: Verification to pass

Instead, this happened: Verification fails with 
```
SUMMARY:
 ** 2 of 316 failed (5 unreachable)
Failed Checks: This is a placeholder message; Kani doesn't support message formatted at runtime
 File: "/github/home/.rustup/toolchains/nightly-2023-07-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs", line 524, in alloc::raw_vec::capacity_overflow
Failed Checks: memcpy source region readable
 File: "/github/home/.rustup/toolchains/nightly-2023-07-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/intrinsics.rs", line 2680, in std::intrinsics::copy_nonoverlapping::<u8>

VERIFICATION:- FAILED
Verification Time: 0.6225813s
```

Using `--visualize` seems to indicate that kani might consider the length of the `const` slice to be not 2, but some very high number (maybe non-deterministic?), which causes an overflow check in the Vec allocation routine to fail (e.g. `count = 9223372036854776320ul (10000000 00000000 00000000 00000000 00000000 00000000 00000010 00000000)`). 

Note that the `memcpy` failure can also be replicated with a separate harness that simply does `MY_CONSTANT.to_vec()`.

**Comment by zhassan-aws** [comment]
Thanks @roypat for reporting this issue!

This seems to happen without involving "strings" as well, e.g.
```rust
#[kani::proof]
fn verify_my_test() {
    const MY_CONSTANT: &[u8] = &[0];

    let mut my_vec: Vec<u8> = Vec::new();
    my_vec.extend(MY_CONSTANT);
}
```

**Comment by zhassan-aws** [comment]
The issue seems to be a bug in Kani's handling of constants: if I replace `const` with `let`, the test verifies:
```rust
#[kani::proof]
fn verify_my_test() {
    let MY_CONSTANT: &[u8] = &[0];

    let mut my_vec: Vec<u8> = Vec::new();
    my_vec.extend(MY_CONSTANT);
}
```
```

SUMMARY:
 ** 0 of 316 failed (5 unreachable)

VERIFICATION:- SUCCESSFUL
Verification Time: 0.42594692s
```

**Comment by zhassan-aws** [comment]
I'm getting failures without using`Vec` as well, e.g.:
```rust
use std::convert::TryInto;

#[kani::proof]
fn verify_my_test() {
    const MY_CONSTANT: &[u8] = &[0];
    let x: [u8; 1] = MY_CONSTANT.try_into().unwrap();

}
```
```
SUMMARY:
 ** 1 of 30 failed (1 unreachable)
Failed Checks: dereference failure: pointer outside object bounds
 File: "/home/ubuntu/.rustup/toolchains/nightly-2023-07-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/array/mod.rs", line 209, in std::array::<impl std::convert::TryFrom<&[u8]> for [u8; 1]>::try_from::{closure#0}
```

**Comment by zhassan-aws** [comment]
This smaller test fails as well:
```rust
#[kani::proof]
fn main() {
    const MY_CONSTANT: &[u8] = &[0];
    let _x: u8 = MY_CONSTANT[0];

}
```
```
SUMMARY:
 ** 1 of 7 failed
Failed Checks: dereference failure: pointer outside object bounds
 File: "/home/ubuntu/examples/iss2656/test.rs", line 4, in main

VERIFICATION:- FAILED
```
